File: Python/Restore IP Addresses/Restore IP Addresses.py
class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        n,ret=len(s),[]
        def dfs(m:List[str],start:int):
            if len(m)==4:
                if start==n:
                    ret.append('.'.join(m))
                return
            for i in range(1,4):
                if i+start>n:
                    return
                if s[start]=='0' and i>1:
                    return
                if int(s[start:i+start])>255:
                    return
                m.append(s[start:i+start])
                dfs(m,start+i)
                m.pop()
        dfs([],0)
        return ret

    

File: C#/Restore IP Addresses/Restore IP Addresses.cs
public class Solution {
    public IList<string> RestoreIpAddresses(string s) {
        IList<string> res=new List<string>();
        IList<string> tmp=new List<string>();
        DFS(ref s,ref res,ref tmp,0);
        return res;
    }

    public void DFS(ref string s,ref IList<string> res,ref IList<string> tmp,int startIdx){
        if(tmp.Count==4){
            if(startIdx==s.Length) res.Add(string.Join('.',tmp));
            return;
        }
        string sub;
        for(int i=1;i<=3;i++){
            if(i+startIdx>s.Length) return;
            if(i>1 && s[startIdx]=='0') return;
            sub=s.Substring(startIdx,i);
            if(i==3 && int.Parse(sub)>255) return;
            tmp.Add(sub);
            DFS(ref s,ref res,ref tmp,startIdx+i);
            tmp.RemoveAt(tmp.Count-1);
        }
    }
}